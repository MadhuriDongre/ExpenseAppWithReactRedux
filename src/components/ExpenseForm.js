import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css'

const now = moment().format('MMM Do, YYYY');
console.log(now)

export default class ExpenseForm extends React.Component {
    constructor(props){
        super();
        this.state={
            description:props.expense?props.expense.description:'',
            note:props.expense?props.expense.note:'',
            amount:props.expense?props.expense.amount.toString():'',
            createdAt: props.expense?moment(props.expense.createdAt):moment(),
            focused:false,
            error:null
        };
    }
    onDescriptionChange=(e)=>{
        const description = e.target.value;
        this.setState(()=>({description}));
    };

    onNoteChange=(e)=>{
        const note = e.target.value;
        this.setState(()=>({note}));
    };
    onAmountChange=(e)=>{
        const amount = e.target.value;
        (!amount ||amount.match(/^\d{1,}(\.\d{0,2})?$/))? this.setState(()=>({amount})):null;
    }
    onDateChange=(createdAt)=>{
        // const createdAt=e.target.value
        createdAt?this.setState(()=>({createdAt})):null
    }
    onFocusChange=({focused})=>{
        this.setState(()=>({focused}))
    }
    onSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            const error= "Please provide Description and Amount to create expense"
            this.setState(()=>({error}))
        }
        else{
            this.setState(()=>({error:null}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10),
                note:this.state.note,
                createdAt:this.state.createdAt.valueOf()
            })
        }
    }
    render(){
        return (
            <div>
                <p>ExpenseForm text</p>
                <form onSubmit={this.onSubmit}> 
                    <input type="text" placeholder="description" value={this.state.description} autoFocus
                        onChange={this.onDescriptionChange}
                    />
                    <input type="text" value ={this.state.amount} placeholder="amount"
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.focused} // PropTypes.bool
                        onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                    />
                    <textarea placeholder="add a note for expense (optional)" value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button>Add Expense</button>
                    {this.state.error && <p className="warning">{this.state.error}</p>}
                </form>
            </div>
        );
    }
}