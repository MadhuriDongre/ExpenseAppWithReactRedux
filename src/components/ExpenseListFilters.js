import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate,setEndDate,setStartDate} from '../actions/filters';
import { DateRangePicker } from 'react-dates'

class ExpenseListFilters extends React.Component{
    state={
        focused:null
    }
    onDatesChange=({startDate, endDate})=>{
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }
    onFocusChange=(focused)=>{
        this.setState(()=>({focused}))
    }   
    render(){
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e)=>{
                    this.props.dispatch(setTextFilter(e.target.value));
                }}/>
                <select value={this.props.filters.sortBy} onChange={(e)=>{
                    e.target.value==="amount"?this.props.dispatch(sortByAmount(e.target.value)):this.props.dispatch(sortByDate(e.target.value));
                }}>
                    <option value ="date">Date</option>
                    <option value ="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                />
            </div>
        )
    }
}
// const ExpenseListFilters =(props)=>(
//     <div>
//         <input type="text" value={props.filters.text} onChange={(e)=>{
//             props.dispatch(setTextFilter(e.target.value));
//         }}/>
//         <select value={props.filters.sortBy} onChange={(e)=>{
//             e.target.value==="amount"?props.dispatch(sortByAmount(e.target.value)):props.dispatch(sortByDate(e.target.value));
//         }}>
//             <option value ="date">Date</option>
//             <option value ="amount">Amount</option>
//         </select>

//     </div>
// );

const mapStateToProps=(state)=>{
    return {
        filters:state.filters
    }
}
export default connect(mapStateToProps)(ExpenseListFilters);