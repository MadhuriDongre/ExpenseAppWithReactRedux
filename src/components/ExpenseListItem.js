import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral'

const ExpenseListItem =({description,amount,createdAt,id})=>(
    <div>
        <h2><Link to={`/edit/${id}`}>{description}</Link></h2>
        <p> {<span>&#x20B9;</span>}{ numeral(amount).format('0,0[.]00') } -- {moment(createdAt).format('MMMM Do, YYYY')}</p>
    </div>
);

export default ExpenseListItem;