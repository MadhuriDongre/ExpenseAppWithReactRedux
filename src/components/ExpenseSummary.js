import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral'
import selectExpenseTotal from '../selectors/expenses-total';
import selectExpense from '../selectors/expenses';

export const ExpenseSummary =({expenseCount, expensesTotal})=>(
    <div>
        <p>{`Viewing ${expenseCount} expense(s) totalling `}{<span>&#x20B9;</span>}{` ${numeral(expensesTotal).format('0,0[.]00')}`}</p>
    </div>
);

const mapStateToProps = (state)=>{
    const visibleExpense=selectExpense(state.expenses,state.filters)
    return {
        expenseCount:visibleExpense.length,
        expensesTotal:selectExpenseTotal(visibleExpense)
    }
};

export default connect(mapStateToProps)(ExpenseSummary);