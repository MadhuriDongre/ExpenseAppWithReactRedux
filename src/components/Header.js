import React from 'react';
import { NavLink} from 'react-router-dom';
import { connect} from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header =({startLogout})=>(
    <header>
        <h1>Expensify</h1>
            <NavLink activeClassName="isActive" to="/dashboard" exact>Dashboard</NavLink>
            <NavLink activeClassName="isActive" to="/create">Create Expense</NavLink>
            <button onClick={startLogout}>LogOut</button>
    </header>
);

const mapDispatchToProps=(dispatch)=>({
    startLogout:()=>dispatch(startLogout())
})
export default connect(undefined,mapDispatchToProps)(Header);