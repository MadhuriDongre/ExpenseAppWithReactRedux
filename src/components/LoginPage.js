import React , { useState }from 'react';
import { connect} from 'react-redux';
import { startLoginWithGoogle, startLoginWithFacebook,startLoginWithTwitter, startLoginWithEmail , startSignupWithEmail, resetPassword} from '../actions/auth';

export const LoginPage = ({startLoginWithGoogle, startLoginWithFacebook, startLoginWithTwitter, startLoginWithEmail, startSignupWithEmail , resetPassword}) =>{
    const [email,setEmail] = useState('');
    const [password,setPassword] = useState('');
    const [password1,setPassword1] = useState('');
    const [error, setError]= useState('');
    const [message, setMessage]= useState('');
    const [showSignupPage, setShowSignupPage] = useState(false);

    const resetPasswordCall =()=>{
        if(email){
            resetPassword(email)
                .then(()=>setMessage('To reset Password, click the link provide in your registered Email Id'))
                .catch((err)=>setError(err))
        }
    }

    const loginWithEmail = (e)=>{
        e.preventDefault();
        if(!email){
            setError(`Email cannot be empty`);
        }
        else if(!password){
            setError(`Password cannot be empty`);
        }
        else{
            startLoginWithEmail( email, password )
            .then(()=>console.log('Login Successful!!'))
            .catch((err)=>{
                console.log('Login Failed!!!', err.code);
                setError(`Please provide correct Email/Password`);
            })
        }
    }

    const signUpWithEmail = (e)=>{
        e.preventDefault();
        if(!email){
            setError(`Email cannot be empty`);
        }
        else if(password!==password1){
            setError(`Passwords don't match`);
        }
        else{
            startSignupWithEmail( email, password )
        }
    }

    return (
        <div className="box-layout">
            <div className="box-layout__box">
                <h1 className="box-layout__title">Expensify</h1>
                <p>It's time to get your expenses under control! </p>
                {!showSignupPage && 
                    <div>
                        <form className="form" onSubmit={ loginWithEmail } >
                            {error && <p className="form__error">{error}</p>}
                            {message && <p className="form__message">{message}</p>}
                            <input className="text-input text-input--login" type="text" name="email" value={email} onChange={(e)=>setEmail(e.target.value)} placeholder="Email Address"/>
                            <input className="text-input text-input--login" type="password" name="password" value={password} onChange={(e)=>setPassword(e.target.value)} placeholder="Password"/>
                            <button className="button button--link button--login" type="submit">Log In</button>
                        </form>
                        <p className="form__text">Forgot Password? <button className="button_mini" onClick={resetPasswordCall}>Reset Password</button></p>
                        <p className="form__text">Don't have an account? <button className="button_mini" onClick={()=>setShowSignupPage(true)}>Sign up</button></p>
                        <p className="form__text">OR</p>
                        <p className="form__text">Login With</p>
                        <button className="button button--link" onClick={startLoginWithGoogle}><img src={'assets/icons8-google.svg'} alt="Logo" /></button>
                        <button className="button button--link" onClick={startLoginWithFacebook}><img src={'assets/icons8-facebook.svg'} alt="Logo" /></button>
                        <button className="button button--link" onClick={startLoginWithTwitter}><img src={'assets/icons8-twitter.svg'} alt="Logo" /></button>
                    </div>
                }
                {showSignupPage && 
                    <div>
                        <p>Register yourself</p>
                        <form className="form" onSubmit={ signUpWithEmail }>
                            {error && <p className="form__error">{error}</p>}
                            <input className="text-input text-input--login" type="text" name="email" value={email} onChange={(e)=>setEmail(e.target.value)} placeholder="Email Address"/>
                            <input className="text-input text-input--login" type="password" name="password" value={password} onChange={(e)=>setPassword(e.target.value)} placeholder="Password"/>
                            <input className="text-input text-input--login" type="text" name="password" value={password1} onChange={(e)=>setPassword1(e.target.value)} placeholder="Confirm Password"/>
                            <button className="button button--link button--login" type="submit">Create Account</button>
                        </form>
                    </div>
                }
            </div>
        </div>
    )
}

const mapDispatchToProps=(dispatch)=>({
    startLoginWithGoogle:()=>dispatch(startLoginWithGoogle()),
    startLoginWithFacebook:()=>dispatch(startLoginWithFacebook()),
    startLoginWithTwitter:()=>dispatch(startLoginWithTwitter()),
    startLoginWithEmail:(email,password)=>dispatch(startLoginWithEmail(email, password)),
    startSignupWithEmail:(email,password)=>dispatch(startSignupWithEmail(email, password)),
    resetPassword:(email)=>dispatch(resetPassword(email))
});

export default connect(undefined,mapDispatchToProps)(LoginPage);