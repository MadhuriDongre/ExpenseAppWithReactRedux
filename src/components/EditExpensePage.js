import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense} from '../actions/expenses'

const EditExpensePage =(props)=>{
    console.log(props);
    return (
        <div>
           <ExpenseForm 
                onSubmit={(expense)=>{
                    props.dispatch(editExpense(props.expenses.id,expense));
                    props.history.push('/');
                }}
                expense={props.expenses}
            />
            <button onClick={()=>{
                props.dispatch(removeExpense({id:props.expenses.id}));
                props.history.push('/')
            }}>Remove</button>
        </div>
    );
};

const mapStateToProps=(state,props)=>{
    return {
        expenses:state.expenses.find((expense)=>expense.id === props.match.params.id)
    }
}
export default connect(mapStateToProps)(EditExpensePage);