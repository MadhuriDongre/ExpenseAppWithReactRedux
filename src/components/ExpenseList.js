import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpense from '../selectors/expenses';

export const ExpenseList =(props)=>(
    <div>
        <p>ExpenseList</p>
        {props.expenses.length===0?(<p>No expense found!!</p>): (props.expenses  && props.expenses.map((expense)=>{
            return (
                <ExpenseListItem key={expense.id} {...expense}/>
            )}))
        }
    </div>
);

const mapStateToProps = (state)=>{
    return {
        expenses:selectExpense(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);