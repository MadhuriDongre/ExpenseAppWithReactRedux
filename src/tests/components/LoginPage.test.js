import React from 'react';
import { shallow } from 'enzyme';
import  { LoginPage }  from '../../components/LoginPage';

let startLoginWithGoogle,startLoginWithFacebook,startLoginWithTwitter,wrapper;
beforeEach(()=>{
    startLoginWithGoogle = jest.fn()
    startLoginWithFacebook = jest.fn()
    startLoginWithTwitter = jest.fn()
    wrapper=shallow(<LoginPage 
        startLoginWithGoogle={startLoginWithGoogle} 
        startLoginWithFacebook={startLoginWithFacebook}
        startLoginWithTwitter={startLoginWithTwitter}
        />)
})
test('should render Login page correctly',()=>{
    expect(wrapper).toMatchSnapshot();
});

test('should call startLoginWithGoogle on button click',()=>{
    wrapper.find('button').at(3).simulate('click');
    expect(startLoginWithGoogle).toHaveBeenCalled();
});

test('should call startLoginWithFacebook on button click',()=>{
    wrapper.find('button').at(4).simulate('click');
    expect(startLoginWithFacebook).toHaveBeenCalled();
});

test('should call startLoginWithTwitter on button click',()=>{
    wrapper.find('button').at(5).simulate('click');
    expect(startLoginWithTwitter).toHaveBeenCalled();
});
