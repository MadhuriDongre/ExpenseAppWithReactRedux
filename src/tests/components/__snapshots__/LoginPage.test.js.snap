// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login page correctly 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Expensify
    </h1>
    <p>
      It's time to get your expenses under control! 
    </p>
    <div>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <input
          className="text-input text-input--login"
          name="email"
          onChange={[Function]}
          placeholder="Email Address"
          type="text"
          value=""
        />
        <input
          className="text-input text-input--login"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          className="button button--link button--login"
          type="submit"
        >
          Log In
        </button>
      </form>
      <p
        className="form__text"
      >
        Forgot Password? 
        <button
          className="button_mini"
          onClick={[Function]}
        >
          Reset Password
        </button>
      </p>
      <p
        className="form__text"
      >
        Don't have an account? 
        <button
          className="button_mini"
          onClick={[Function]}
        >
          Sign up
        </button>
      </p>
      <p
        className="form__text"
      >
        OR
      </p>
      <p
        className="form__text"
      >
        Login With
      </p>
      <button
        className="button button--link"
        onClick={[MockFunction]}
      >
        <img
          alt="Logo"
          src="assets/icons8-google.svg"
        />
      </button>
      <button
        className="button button--link"
        onClick={[MockFunction]}
      >
        <img
          alt="Logo"
          src="assets/icons8-facebook.svg"
        />
      </button>
      <button
        className="button button--link"
        onClick={[MockFunction]}
      >
        <img
          alt="Logo"
          src="assets/icons8-twitter.svg"
        />
      </button>
    </div>
  </div>
</div>
`;
